accounts-db/src/accounts.rs

```diff
@@ -552,31 +569,52 @@ impl Accounts {
         account_locks: &mut AccountLocks,
         writable_keys: Vec<&Pubkey>,
         readonly_keys: Vec<&Pubkey>,
-    ) -> Result<()> {
+        entry_accts_lookup: &mut EntryAcctLocks,
+        allow_self_conflicting_entries: bool,
+    ) -> (Result<()>, bool) {
+        let mut self_conflicting_account: bool = false;
+        let mut self_conflicting_tx = false;
         for k in writable_keys.iter() {
             if account_locks.is_locked_write(k) || account_locks.is_locked_readonly(k) {
-                debug!("Writable account in use: {:?}", k);
-                return Err(TransactionError::AccountInUse);
+                if allow_self_conflicting_entries {
+                    self_conflicting_account = entry_accts_lookup.writables.contains(k)
+                        || entry_accts_lookup.readables.contains(k);
+                }
+                if !self_conflicting_account {
```

Comment:
I'm confused by the logic here.
What is `self_conflicting_account` and `self_conflicting_tx` supposed to represent?

if there's a conflict with the global locks (`acount_locks`), then we enter the outer `if`.
if the feature is activated, then we set `self_conflicting_account` to be whether this batch conflicts (assuming previous txs in the batch were added to locks!).
then we check if `!self_conflicting_account`, returning error if so.

I think maybe the separated `if` is just confusing, `self_conflicting_account` should never be true if feature is not activated.

Can we re-write this more clearly?

```rust
if allow_self_conflicting_entries {
    if !(entry_accts_lookup.writables.contains(k) || entry_accts_lookup.readables.contains(k)) {
        return (Err(TransactionError::AccountInUse), false); // I think false here doesn't matter since we conflict with global locks, right?
    }
} else {
    return (Err(TransactionError::AccountInUse), false);
}
```

Reply:
To me this is much more clear.
We have 2 cases that can lead to error if we detect a global locks conflict:
1. feature is activated, but the lock is **not** in the locks for this batch.
2. feature is not activated

Reply:
yup, this is more clearer. `self_conflicting_account` is just an intermediate flag that is redundant.

and yes, if a batch is conflicting with another batch then the `self_conflicting_tx` flag is irrelevant.